### ISSUE TITLE
Google Ads lib error "INTERNAL: Panic! This is a bug!"

### ISSUE THREAD

**Issue by aharbunou** [googleads/google-ads-java#194]
Hi,

We have a service that communicates to Google Ads API via the lib. 
It works fine for some period of time after a deployment (like a week or less). Then we see only "INTERNAL: Panic! This is a bug!" errors for any(!) request we send. We don't see any request/response logs in this case. My guess that exception happened before sending the request. 
Version we use is 4.2.0 but we saw the same thing with 4.0.0.
The issue is kind of critical for us because it starts unpredictably and have immediate impact for clients of our service.

Please let me know if you need more details from our side.

Stacktrace : 

com.google.api.gax.rpc.InternalException: io.grpc.StatusRuntimeException: INTERNAL: Panic! This is a bug!
	at com.google.api.gax.rpc.ApiExceptionFactory.createException(ApiExceptionFactory.java:67)
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:72)
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:60)
	at com.google.api.gax.grpc.GrpcExceptionCallable$ExceptionTransformingFuture.onFailure(GrpcExceptionCallable.java:97)
	at com.google.api.core.ApiFutures$1.onFailure(ApiFutures.java:68)
	at com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1070)
	..........
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:592)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$700(ClientCallImpl.java:508)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:632)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
	Suppressed: com.google.api.gax.rpc.AsyncTaskException: Asynchronous task failed
		at com.google.api.gax.rpc.ApiExceptions.callAndTranslateApiException(ApiExceptions.java:57)
		at com.google.api.gax.rpc.UnaryCallable.call(UnaryCallable.java:112)
		at com.google.ads.googleads.v2.services.GoogleAdsServiceClient.search(GoogleAdsServiceClient.java:247)
		........
		at io.dropwizard.jetty.RoutingHandler.handle(RoutingHandler.java:52)
		at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:455)
		at io.dropwizard.jetty.BiDiGzipHandler.handle(BiDiGzipHandler.java:63)
		at org.eclipse.jetty.server.handler.RequestLogHandler.handle(RequestLogHandler.java:56)
		at org.eclipse.jetty.server.handler.StatisticsHandler.handle(StatisticsHandler.java:169)
		at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
		at org.eclipse.jetty.server.Server.handle(Server.java:530)
		at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:347)
		at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:256)
		at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:279)
		at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:102)
		at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:124)
		at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:247)
		at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produce(EatWhatYouKill.java:140)
		at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131)
		at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:382)
		at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:708)
		at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:626)
		... 1 common frames omitted
Caused by: io.grpc.StatusRuntimeException: INTERNAL: Panic! This is a bug!
	at io.grpc.Status.asRuntimeException(Status.java:533)
	... 27 common frames omitted
Caused by: java.util.concurrent.RejectedExecutionException: event executor terminated
	at io.grpc.netty.shaded.io.netty.util.concurrent.SingleThreadEventExecutor.reject(SingleThreadEventExecutor.java:845)
	at io.grpc.netty.shaded.io.netty.util.concurrent.SingleThreadEventExecutor.offerTask(SingleThreadEventExecutor.java:328)
	at io.grpc.netty.shaded.io.netty.util.concurrent.SingleThreadEventExecutor.addTask(SingleThreadEventExecutor.java:321)
	at io.grpc.netty.shaded.io.netty.util.concurrent.SingleThreadEventExecutor.execute(SingleThreadEventExecutor.java:756)
	at io.grpc.netty.shaded.io.netty.util.concurrent.AbstractScheduledEventExecutor.schedule(AbstractScheduledEventExecutor.java:232)
	at io.grpc.netty.shaded.io.netty.util.concurrent.AbstractScheduledEventExecutor.schedule(AbstractScheduledEventExecutor.java:155)
	at io.grpc.netty.shaded.io.netty.util.concurrent.AbstractEventExecutorGroup.schedule(AbstractEventExecutorGroup.java:50)
	......
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:626)

**UPDATE :** 

I also noticed the following error right before an instance started "panicking". Not sure whether it is related or not.

i.g.i.LogExceptionRunnable -  Exception while executing runnable io.grpc.internal.SharedResourceHolder$2@1b39c0af 
io.grpc.netty.shaded.io.netty.channel.ChannelException: eventfd_write() failed: Bad file descriptor
	at io.grpc.netty.shaded.io.netty.channel.epoll.Native.eventFdWrite(Native Method)
	at io.grpc.netty.shaded.io.netty.channel.epoll.EpollEventLoop.wakeup(EpollEventLoop.java:167)
	at io.grpc.netty.shaded.io.netty.util.concurrent.SingleThreadEventExecutor.shutdownGracefully(SingleThreadEventExecutor.java:603)
	at io.grpc.netty.shaded.io.netty.util.concurrent.MultithreadEventExecutorGroup.shutdownGracefully(MultithreadEventExecutorGroup.java:163)
	at io.grpc.netty.shaded.io.grpc.netty.Utils$DefaultEventLoopGroupResource.close(Utils.java:346)
	at io.grpc.netty.shaded.io.grpc.netty.Utils$DefaultEventLoopGroupResource.close(Utils.java:318)
	at io.grpc.internal.SharedResourceHolder$2.run(SharedResourceHolder.java:145)
	at io.grpc.internal.LogExceptionRunnable.run(LogExceptionRunnable.java:43)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

**Comment by jradcliff** [comment]
Hi,

Could you provide a bit more detail about how you are using instances of service clients? For example, do you call [`createGoogleAdsServiceClient()`](https://github.com/googleads/google-ads-java/blob/fa524016543a7d8a3008b275b5dff950884b1338/google-ads/src/main/java/com/google/ads/googleads/v2/services/GoogleAdsVersion.java#L220) once at startup and then hang onto and reuse that instance of `GoogleAdsServiceClient` for hours or days?

I ask because one possibility is that the gRPC channel associated with that particular long-lived instance of `GoogleAdsServiceClient` is somehow getting in a bad state. If that's the case, you could periodically get a new instance of that class from the factory to see if that helps.

Thanks,
Josh, Google Ads API Team

**Comment by aharbunou** [comment]
Hi @jradcliff ,

Thank you for looking on this!
We call ` createGoogleAdsServiceClient() ` every time before search. See code in Kotlin below 
```
googleAdsClient
.version2
.createGoogleAdsServiceClient()
.use { client -> return client.search(searchRequest).iterateAll().toList() } 
```

However, we create only one instance of ` GoogleAdsClient ` class. 
Can this cause the issue?

Thanks,
Andrei

**Comment by jradcliff** [comment]
Hi Andrei,

There are some attributes of `GoogleAdsClient` that will be reused throughout the lifecycle of an instance, so it's possible one of those is winding up in a bad state.

Could you try occasionally fetching a new instance of `GoogleAdsClient` and let me know if that fixes the issue? If so, that'll be very useful info we can use to investigate further.

A potentially interesting note is that the source of the message you're seeing appears to be [this line in ManagedChannelImpl.java](https://github.com/grpc/grpc-java/blob/eb21c646b239fdf00e25abf5eaa3b7b523f5cf05/core/src/main/java/io/grpc/internal/ManagedChannelImpl.java#L797), from the [grpc-java](https://github.com/grpc/grpc-java) dependency.

Best regards,
Josh, Google Ads API Team

**Comment by aharbunou** [comment]
Thank you @jradcliff for the suggestion. 
We will definitely try it and let you know if we notice the issue again.

Thanks,
Andrei

**Comment by aharbunou** [comment]
Hi @jradcliff. 

We did the proposed change, but unfortunately the issue happened again today.
Also, I've found 
``` io.grpc.netty.shaded.io.netty.channel.ChannelException: eventfd_write() failed: Bad file descriptor ``` 
error again right before "panic" started. 
Could you please help to fix it?

Thanks,
Andrei

**Comment by jradcliff** [comment]
Hi Andrei,

I looked at [related issues in the grpc-java project](https://github.com/grpc/grpc-java/issues?utf8=%E2%9C%93&q=is%3Aissue++panic+), and they suggest this could be due to a resource issue for your process. Do you see any OutOfMemory errors or any other errors of note (in addition to the netty error you highlighted) before this error occurs?

Thanks,
Josh Radcliff, Google Ads API Team

**Comment by aharbunou** [comment]
@jradcliff memory was the first thing I checked before the issue creation.
Unfortunately (as it would be a quick win :) ), I don't see any OOM for last weeks and our JVM metrics don't show spikes in CPU or memory. File descriptor error is the only weird thing.

Thanks,
Andrei

**Comment by jradcliff** [comment]
Hi Andrei,

Sorry for all the questions, but I want to try to produce the error on my side so need a bit more information.

When you create a `GoogleAdsClient`, do you reuse it across multiple threads? If so, what's the max number of threads you have running at any given time?

Thanks,
Josh

**Comment by aharbunou** [comment]
@jradcliff it is OK. I will provide whatever details you need if it helps :) 

1. After your last comment I've updated the code so that we create new instance of ` GoogleAdsClient ` per each call to Google API to avoid issues with sharing the object between threads. I also didn't notice any performance or memory issues after the change. 
Simple performance test shows that first ` GoogleAdsClient ` instance creation takes around 1 sec but then it takes milliseconds to create any subsequent one. Is there any static cache in the library?

2. Yep, we do close. Our codebase is on Kotlin so we do the following (for example campaign mutation call)
```
....
.createCampaignServiceClient()
                        .use { client ->
                            client.mutateCampaigns(accountId, operations)
                        }
...
```
`use` in Kotlin is equivalent of `try-with-resource` in Java.

Thanks,
Andrei

**Comment by nwbirnie** [comment]
Looks like the only place that this panic can be triggered from is [here](https://github.com/grpc/grpc-java/blob/75f6fd8f105d26d1a1389fd2d735deee8b4234e1/core/src/main/java/io/grpc/internal/ManagedChannelImpl.java#L150-L160). It's an uncaught exception on the [gRPC event dispatch queue](https://github.com/grpc/grpc-java/blob/75f6fd8f105d26d1a1389fd2d735deee8b4234e1/api/src/main/java/io/grpc/SynchronizationContext.java#L93-L99).

@aharbunou can you enable the gRPC logs please? They're logged to JUL (java util logging) so creating a configuration file like the following and setting `-Djava.util.logging.config.file=jul.properties` should send these to the console.

```
handlers= java.util.logging.ConsoleHandler
.level=ALL
java.util.logging.ConsoleHandler.level=ALL
java.util.logging.ConsoleHandler.formatter=java.util.logging.SimpleFormatter
```

We're looking for the exception which is logged with the message "Uncaught exception in the SynchronizationContext. Panic!". Thanks.

**Comment by aharbunou** [comment]
@nwbirnie we already have this exception in our log. Please find one example below. I also checked that *all* of them have the same "RejectedExecutionException" inside.
```
[Channel<1702603>: (googleads.googleapis.com:443)] Uncaught exception in the SynchronizationContext. Panic! 
java.util.concurrent.RejectedExecutionException: event executor terminated
  at io.grpc.netty.shaded.io.netty.util.concurrent.SingleThreadEventExecutor.reject(SingleThreadEventExecutor.java:845)
  at io.grpc.netty.shaded.io.netty.util.concurrent.SingleThreadEventExecutor.offerTask(SingleThreadEventExecutor.java:328)
  at io.grpc.netty.shaded.io.netty.util.concurrent.SingleThreadEventExecutor.addTask(SingleThreadEventExecutor.java:321)
  at io.grpc.netty.shaded.io.netty.util.concurrent.SingleThreadEventExecutor.execute(SingleThreadEventExecutor.java:756)
  at io.grpc.netty.shaded.io.netty.util.concurrent.AbstractScheduledEventExecutor.schedule(AbstractScheduledEventExecutor.java:232)
  at io.grpc.netty.shaded.io.netty.util.concurrent.AbstractScheduledEventExecutor.schedule(AbstractScheduledEventExecutor.java:155)
  at io.grpc.netty.shaded.io.netty.util.concurrent.AbstractEventExecutorGroup.schedule(AbstractEventExecutorGroup.java:50)
  at io.grpc.netty.shaded.io.netty.util.concurrent.AbstractEventExecutorGroup.schedule(AbstractEventExecutorGroup.java:32)
  at io.grpc.internal.Rescheduler.reschedule(Rescheduler.java:63)
  at io.grpc.internal.ManagedChannelImpl.rescheduleIdleTimer(ManagedChannelImpl.java:408)
  at io.grpc.internal.ManagedChannelImpl.exitIdleMode(ManagedChannelImpl.java:366)
  at io.grpc.internal.ManagedChannelImpl$ChannelTransportProvider$1ExitIdleModeForTransport.run(ManagedChannelImpl.java:467)
  at io.grpc.SynchronizationContext.drain(SynchronizationContext.java:88)
  at io.grpc.SynchronizationContext.execute(SynchronizationContext.java:120)
  at io.grpc.internal.ManagedChannelImpl$ChannelTransportProvider.get(ManagedChannelImpl.java:471)
  at io.grpc.internal.ClientCallImpl.startInternal(ClientCallImpl.java:255)
  at io.grpc.internal.ClientCallImpl.start(ClientCallImpl.java:188)
  at io.grpc.internal.CensusTracingModule$TracingClientInterceptor$1.start(CensusTracingModule.java:394)
  at io.grpc.internal.CensusStatsModule$StatsClientInterceptor$1.start(CensusStatsModule.java:695)
  at io.grpc.ForwardingClientCall.start(ForwardingClientCall.java:32)
  at com.google.api.gax.grpc.GrpcHeaderInterceptor$1.start(GrpcHeaderInterceptor.java:94)
  at io.grpc.ForwardingClientCall.start(ForwardingClientCall.java:32)
  at com.google.ads.googleads.lib.logging.LoggingInterceptor$1.start(LoggingInterceptor.java:78)
  at io.grpc.stub.ClientCalls.startCall(ClientCalls.java:310)
  at io.grpc.stub.ClientCalls.asyncUnaryRequestCall(ClientCalls.java:282)
  at io.grpc.stub.ClientCalls.futureUnaryCall(ClientCalls.java:191)
  at com.google.api.gax.grpc.GrpcDirectCallable.futureCall(GrpcDirectCallable.java:58)
  at com.google.api.gax.grpc.GrpcUnaryRequestParamCallable.futureCall(GrpcUnaryRequestParamCallable.java:65)
  at com.google.api.gax.grpc.GrpcExceptionCallable.futureCall(GrpcExceptionCallable.java:64)
  at com.google.api.gax.rpc.AttemptCallable.call(AttemptCallable.java:86)
  at com.google.api.gax.rpc.RetryingCallable.futureCall(RetryingCallable.java:63)
  at com.google.api.gax.rpc.RetryingCallable.futureCall(RetryingCallable.java:41)
  at com.google.ads.googleads.lib.ExceptionTransformingCallable.futureCall(ExceptionTransformingCallable.java:51)
  at com.google.api.gax.rpc.PagedCallable.futureCall(PagedCallable.java:62)
  at com.google.api.gax.rpc.UnaryCallable$1.futureCall(UnaryCallable.java:126)
  at com.google.api.gax.rpc.UnaryCallable.futureCall(UnaryCallable.java:87)
  at com.google.api.gax.rpc.UnaryCallable.call(UnaryCallable.java:112)
  at com.google.ads.googleads.v2.services.GoogleAdsServiceClient.search(GoogleAdsServiceClient.java:247)
  at com.expedia.www.mtt.googleads.service.gha.client.BaseGhaReadOnlyRepo$trySearch$1.invoke(GhaRepo.kt:50)
 .......
  at com.expedia.www.mtt.googleads.service.resources.GoogleAdsUploadResource.process(GoogleAdsUploadResource.kt:54)
  at sun.reflect.GeneratedMethodAccessor468.invoke(Unknown Source)
  at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  at java.lang.reflect.Method.invoke(Method.java:498)
  at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory$1.invoke(ResourceMethodInvocationHandlerFactory.java:81)
  at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:144)
  at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:161)
  at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:160)
  at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:99)
  at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:389)
  at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:347)
  at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:102)
  at org.glassfish.jersey.server.ServerRuntime$2.run(ServerRuntime.java:326)
  at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271)
  at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267)
  at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
  at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
  at org.glassfish.jersey.internal.Errors.process(Errors.java:267)
  at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:317)
  at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:305)
  at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:1154)
  at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:473)
  at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:427)
  at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:388)
  at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:341)
  at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:228)
  at io.dropwizard.jetty.NonblockingServletHolder.handle(NonblockingServletHolder.java:49)
  at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1650)
  at io.dropwizard.servlets.ThreadNameFilter.doFilter(ThreadNameFilter.java:35)
  at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1637)
  at io.dropwizard.jersey.filter.AllowedMethodsFilter.handle(AllowedMethodsFilter.java:45)
  at io.dropwizard.jersey.filter.AllowedMethodsFilter.doFilter(AllowedMethodsFilter.java:39)
  at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1637)
  at ch.qos.logback.access.servlet.TeeFilter.doFilter(TeeFilter.java:53)
  at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1637)
  at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:533)
  at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:188)
  at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1253)
  at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:168)
  at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:473)
  at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:166)
  at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1155)
  at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
  at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
  at com.codahale.metrics.jetty9.InstrumentedHandler.handle(InstrumentedHandler.java:239)
  at io.dropwizard.jetty.RoutingHandler.handle(RoutingHandler.java:52)
  at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:455)
  at io.dropwizard.jetty.BiDiGzipHandler.handle(BiDiGzipHandler.java:63)
  at org.eclipse.jetty.server.handler.RequestLogHandler.handle(RequestLogHandler.java:56)
  at org.eclipse.jetty.server.handler.StatisticsHandler.handle(StatisticsHandler.java:169)
  at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
  at org.eclipse.jetty.server.Server.handle(Server.java:530)
  at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:347)
  at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:256)
  at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:279)
  at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:102)
  at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:124)
  at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:247)
  at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produce(EatWhatYouKill.java:140)
  at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131)
  at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:382)
  at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:708)
  at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:626)
  at java.lang.Thread.run(Thread.java:748)
```

**Comment by nwbirnie** [comment]
@aharbunou The main reason I can see for this would be failing to create a new thread having been idle for some time. Are you overriding any of the thread pool settings on GoogleAdsClient.setTransportChannelProvider()?

**Comment by aharbunou** [comment]
@nwbirnie no. I've double checked to code. we don't override any pool settings.
We don't see any spikes in memory or CPU. 
Does the lib require additional free hard drive space?

**Comment by nwbirnie** [comment]
@aharbunou gRPC team advised [here](https://github.com/grpc/grpc-java/issues/6590#issuecomment-572817176) the issue you're seeing could be related to this [bug](https://github.com/grpc/grpc-java/issues/6002). It's been fixed in 1.24.0. 

I've opened this [PR](https://github.com/googleads/google-ads-java/pull/217) to update the gRPC dependencies on our master branch, will likely be a couple days before it's reviewed and released.

**Comment by nwbirnie** [comment]
I've just released v4.4.0 which should resolve this issue. Closing just now, but please reopen the bug if it doesn't solve your issue.

Thanks again for reporting.

**Comment by Ashirkhan857** [comment]
Hey, i faced this issue frequently when i download client records.
GRPC Error io. grpc.StatusRuntimeException: INTERNAL:Internal error encountered.Is there any proper way to get rid of this error.
jar i'm using gax-2.6.1.jar , gax-grpc-2.6.1.jar
